External/FEXCore/Source/Interface/HLE/Thunks/Thunks.cpp

```diff
@@ -156,6 +234,125 @@ namespace FEXCore {
             }
         }
 
+        static void IsLibLoaded(void* ArgsRV) {
+            struct ArgsRV_t {
+                const char *Name;
+                bool rv;     // Guest function to call when branching to original_callee
+            };
+
+            auto &[Name, rv] = *reinterpret_cast<ArgsRV_t*>(ArgsRV);
+
+            auto CTX = Thread->CTX;
+            auto That = reinterpret_cast<ThunkHandler_impl*>(CTX->ThunkHandler.get());
+
+            {
+                std::shared_lock lk(That->ThunksMutex);
+                rv = That->Libs.contains(Name);
+            }
+        }
+
+        static void HostTrampolineForGuestcall(void* ArgsRV) {
+          struct ArgsRV_t {
+              IR::SHA256Sum *HostPacker;
+              uintptr_t GuestUnpacker;
+              uintptr_t GuestTarget;
+              uintptr_t rv;
+          };
+
+          auto &[HostPacker, GuestTarget, GuestUnpacker, rv] = *reinterpret_cast<ArgsRV_t*>(ArgsRV);
+
+          auto const CTX = Thread->CTX;
+          auto const That = reinterpret_cast<ThunkHandler_impl *>(CTX->ThunkHandler.get());
+
+          const GuestcallInfo gci = { GuestUnpacker, GuestTarget };
+
+          // Try first with shared_lock
+          {
+            std::shared_lock lk(That->ThunksMutex);
+
+            auto found = That->GuestcallToHostTrampoline.find(gci);
+            if (found != That->GuestcallToHostTrampoline.end()) {
+              rv = found->second;
+              return;
+            }
+          }
+
+          std::lock_guard lk(That->ThunksMutex);
+
+          // retry lookup with full lock before making a new trampoline to avoid double trampolines
+          {
+            auto found = That->GuestcallToHostTrampoline.find(gci);
+            if (found != That->GuestcallToHostTrampoline.end()) {
+              rv = found->second;
+              return;
+            }
+          }
+
+          auto HostPackerEntry = That->Thunks.find(*HostPacker);
+          if (HostPackerEntry == That->Thunks.end()) {
+            rv = 0;
```

Comment:
This should be a fatal error AFAICT.

Reply:
Not necessarily - it is up to the called to decide what to do there. Ie, a function may return a non-trampoline host pointer registered via some thunked library, and might chose to register a GCH for it.